### DISCUSSION TITLE
Move batch compiler to Java 17, drop Java 11 runtime support

### DISCUSSION THREAD

**Discussion by iloveeclipse** [eclipse-jdt/eclipse.jdt.core#888] [Polls]
See https://github.com/eclipse-jdt/eclipse.jdt.core/issues/886

The proposal is that with 4.28 release Eclipse Java Compiler (ECJ) will drop support for execution on Java **runtime** below 17, and will require **at least Java 17 runtime** for execution.

**The reasons for switch to Java 17 runtime are**:

- **There are no tests running on Java 11 anymore**. So we could have already broken ecj because of some runtime issue we don't see on Java 17 / 19.
- Use latest Java language features that simplify our life: records, instanceof pattern, etc in compiler code itself
- Use latest available JDK libraries, example: [this one line](https://github.com/eclipse-pde/eclipse.pde/blob/8893c7e677e8fb4848e2e75c38e2e259810652e0/ui/org.eclipse.pde.ui.tests/src/org/eclipse/pde/ui/tests/project/ProjectCreationTests.java#L1486) **using Java 17**  was a [complex method with 40 lines](https://github.com/eclipse-pde/eclipse.pde/blob/eaf462c977ca3a1f2e1ef4226f0ba17ec9c8cdab/ui/org.eclipse.pde.ui.tests/src/org/eclipse/pde/ui/tests/project/ProjectCreationTests.java#L1486) in Java 8!
- Simplify our life maintaining compiler source code configuration, build, test etc - all the jenkins/maven/github/ whatever configurations where we should simply **reuse** solutions we would have with default platform improvements we see every day.
- basically: **save our time**. 
- Our development work would only benefit from Java 17.
- Consumers that need ecj running on Java 11 can use older ECJ versions.

**The reasons for stay on Java 11 runtime are**:
- First and foremost, Java 11 is still an active version that is still being supported by Oracle. Platform moving ahead is a complex case, but there's no such reason to do that in the batch compiler case.
- We should take a moment to think about all the ECJ consumers we will either be leaving behind or upsetting. Case in point, the comment from [kriegaex](https://github.com/kriegaex) - https://github.com/eclipse-jdt/eclipse.jdt.core/issues/886#issuecomment-1479638779. Surely there would be other bigger consumers such as Apache Tomcat.
- Java 17 is fairly new, by Java standards and only a year and half old. Those clients that prefer to go with tried and tested software will not be happy about this.

**Important notes**
- Compilation **targets** will be **not** affected, ECJ will still support compilation targets from Java 20 down to Java 1.8.
- Voting deadline is **one week**, means 29.03.2023 is the last voting day.

**Comment by akurtakov** [comment]
Please set a deadline for voting. IMHO 1 week is enough for whoever is interested to cast his/her vote.

**Reply by iloveeclipse** [reply]
> Please set a deadline for voting.

**How**? I don't see any "deadline" or whatever option in the github UI. Or do you mean "add a note to the poll text about deadline"?

**Reply by akurtakov** [reply]
Add a note to the poll text. Such functionality is not yet implemented but the request for it is https://github.com/community/community/discussions/44890

**Reply by iloveeclipse** [reply]
Done

**Comment by jarthana** [comment]
I think it's only fair that we present both sides of the argument. In fact, this should go to the comment 0.

- First and foremost, Java 11 is still an active version that is still being supported by Oracle. Platform moving ahead is a complex case, but there's no such reason to do that in the batch compiler case.
- We should take a moment to think about all the ECJ consumers we will either be leaving behind or upsetting. Case in point, the comment from [kriegaex](https://github.com/kriegaex) - https://github.com/eclipse-jdt/eclipse.jdt.core/issues/886#issuecomment-1479638779. Surely there would be other bigger consumers such as Apache Tomcat.

**Reply by iloveeclipse** [reply]
Thanks Jay, I've updated the voting text to include your comments 1:1. 
I believe you should be able to update description too, so feel free to do so.

**Comment by iloveeclipse** [comment]
OK, closing this one.
Status: 16 votes, 62% "yes", 31% "no", 6% "don't care".
